### ISSUE TITLE
How do we add cli-style to repos?

### ISSUE THREAD

**Issue by ismay** [dhis2/notes#28]
@varl and I were talking [here](https://github.com/dhis2/d2/pull/200#pullrequestreview-211760094) about how to implement cli-style for a repo. I tried starting from a common commit for the settings-app, adding cli-style and then merging to the five (master and 4 back) versions that we're actively using.

This leads to a bunch of conflicts, and I'm a bit worried that that might actually introduce more problems than it solves. Alternative is to just implement for master, which will make merging to master more difficult for our merge-base workflow.

So I'm wondering what others think we should do. Any perspectives on what the best approach would be? For reference, below two PRs that add cli-style to an app.

- https://github.com/dhis2/d2/pull/200
- https://github.com/dhis2/settings-app/pull/132

**Comment by varl** [comment]
One option is to put in cli-style, but not run it with `-all`, just letting it reformat staged files and take over the code base over time.

**Comment by ismay** [comment]
Would you then put it on all (supported) branches? Or just master.

**Comment by varl** [comment]
I think it makes sense to put it on the active branches. We need to consider the supported branches as "live" ones, which deserve our attention and care.

**Comment by ismay** [comment]
As an example, I've implemented the above approach for the settings-app:

- https://github.com/dhis2/settings-app/pull/134
- https://github.com/dhis2/settings-app/pull/135
- https://github.com/dhis2/settings-app/pull/136
- https://github.com/dhis2/settings-app/pull/137
- https://github.com/dhis2/settings-app/pull/138

Let me know what you think of doing it this way.

**Comment by varl** [comment]
> As an example, I've implemented the above approach for the settings-app:
> 
>     * [dhis2/settings-app#134](https://github.com/dhis2/settings-app/pull/134)
> 
>     * [dhis2/settings-app#135](https://github.com/dhis2/settings-app/pull/135)
> 
>     * [dhis2/settings-app#136](https://github.com/dhis2/settings-app/pull/136)
> 
>     * [dhis2/settings-app#137](https://github.com/dhis2/settings-app/pull/137)
> 
>     * [dhis2/settings-app#138](https://github.com/dhis2/settings-app/pull/138)
> 
> 
> Let me know what you think of doing it this way.

I think that's reasonable, it starts to get sticky when/if we run `d2-style js apply --all` ... Which we _do_ want, generally. :|

**Comment by ismay** [comment]
> Which we do want, generally. :|

Yeah that would be ideal. But with our current backporting requirements I wouldn't know how to do it without causing massive conflicts. I think this way seems like the most reasonable option.

**Comment by varl** [comment]
@amcgee What do you think about incremental adoption of the code style?

**Comment by amcgee** [comment]
@varl sorry missed this one last week.  I actually worry a little bit about incremental adoption because it can obfuscate the diffs in functional changes.  For instance, if I change 10 lines in a 200 line file for a functional change (or even worse in 20 200-line files) and those files haven't been `cli-style`'d yet, then I end up with a many-hundred-line diff.

Ideally the `style js apply` would be in a separate (earlier) commit than the functional change, but that's not how it works with the staged-only default and a pre-commit hook.

For the above reason I think I support a single all-in `style js apply --all` non-functional change before adopting the incremental approach - that way the only changes the pre-commit hook should make are to places that were touched in the current change.  Backporting does make this challenging though - I think you'd have to run `style js apply --all` one every actively-developed branch to enable clean cherry-picked commit backports, but that gets pretty noisy and disruptive.

I also worry somewhat about the assumption that the style application is 100% non-functional.  If there's a bug in `prettier` or something and we just blindly `apply --all` to a huge number of files we could miss accidental functional changes or errors introduced in the initial code-style PR, so it's important to have good tests (unit, smoke, and integration) before migrating a codebase to the new style.  This last point leans back towards incremental adoption, though ideally the style and functional changes would be applied in separate commits.

tldr: I'm conflicted, but I **think** I prefer running `apply --all` over incremental adoption.

**Comment by varl** [comment]
>enable clean cherry-picked commit backports

Please see https://dhis2.github.io/guides/04-git-workflow#why-is-cherry-pick-a-problem-when-used-to-port-code-between-branches. Cherry-picking commits across branches is something we try to avoid. Alternative workflows are documented in that article as well.

**Comment by amcgee** [comment]
@varl yeah, definitely not recommending cherry-picking in general.  I don't *think* there's another option if you've independently reformatted each branch with `style apply --all` and therefore broken your branch mergability though.  That's what I was talking about in that paragraph, sorry if it wasn't clear.

**Comment by varl** [comment]
> @varl yeah, definitely not recommending cherry-picking in general. I don't _think_ there's another option if you've independently reformatted each branch with `style apply --all` and therefore broken your branch mergability though. That's what I was talking about in that paragraph, sorry if it wasn't clear.

Ok, cool. Just erring on the side of caution. 👍 

One option is to use `git format-patch <range>` and `git am *.patch` on the other branches. Since it's just a diff I think it would apply clean if the code is formatted the same.

Need to be verified to be sure...

**Comment by amcgee** [comment]
> git format-patch <range>
Sorcery ;-)

**Comment by varl** [comment]
> > `git format-patch <range>` and `git am *.patch`
> 
> Sorcery!

E-mail patches works pretty well for the Linux kernel..

**Comment by ismay** [comment]
@amcgee Hmm yeah I see what you mean. I find it quite difficult to come up with a good migration strategy. The backporting makes this a lot more complicated than a project which just publishes a single branch. Normally I'd also really prefer prettifying the entire codebase at once.

So but if we foresee problems with the incremental approach we could still revert the changes. What's the concensus on the best approach atm.? Do we need to do some further testing?

**Comment by varl** [comment]
Proposition:

- We reformat all files directly.
- We do not backport style application.
- We accept that backports become harder to automatically apply.
- We think that moving forward on master is more important than trying to drag 5 branches along for the ride.

y/n?

**Comment by ismay** [comment]
I agree 👍

**Comment by amcgee** [comment]
Aye!
